src/v/cluster/scheduling/partition_allocator.cc

```diff
@@ -9,325 +9,206 @@
+      current.end(),
+      std::back_inserter(to_add),
+      [&previous](const model::broker_shard& current_bs) {
+          auto it = std::find(previous.begin(), previous.end(), current_bs);
+          return it == previous.end();
       });
-    auto node_id = std::cbegin(shards)->node_id;
-    auto it = find_node(node_id);
-
-    for (auto const& bs : shards) {
-        if (it == _machines.end()) {
-            // do nothing, node was deleted
-            continue;
-        }
-        // Thanks to shards being sorted we need to do only
-        //  as many lookups as there are brokers
-        if (it->first != bs.node_id) {
-            it = find_node(bs.node_id);
-        }
-        if (it != _machines.end()) {
-            it->second->allocate(bs.shard);
-        }
-    }
-}
-
-partition_allocator::iterator
-partition_allocator::find_node(model::node_id id) {
-    return _machines.find(id);
-}
-
-void partition_allocator::unregister_node(model::node_id id) {
-    auto it = std::find_if(
-      _available_machines.begin(),
-      _available_machines.end(),
-      [id](allocation_node& n) { return n._id == id; });
-
-    if (it != _available_machines.end()) {
-        _available_machines.erase(it);
-    }
-    _machines.erase(id);
-}
-
-void partition_allocator::decommission_node(model::node_id id) {
-    auto it = _machines.find(id);
-    if (it == _machines.end()) {
-        throw std::invalid_argument(
-          fmt::format("machine with id {} not found", id));
-    }
-    // mark as decommissioned
-    it->second->decommission();
 
-    // remove from available machines list
-    if (it->second->_hook.is_linked()) {
-        _available_machines.erase(cil_t::s_iterator_to(*it->second));
-    }
-}
-
-void partition_allocator::recommission_node(model::node_id id) {
-    auto it = _machines.find(id);
-    if (it == _machines.end()) {
-        throw std::invalid_argument(
-          fmt::format("machine with id {} not found", id));
-    }
-    // mark as decommissioned
-    it->second->recommission();
-
-    // add to available machines list
-    _available_machines.push_back(*it->second);
-}
-
-bool partition_allocator::is_empty(model::node_id id) {
-    auto it = _machines.find(id);
-    if (it == _machines.end()) {
-        throw std::invalid_argument(
-          fmt::format("machine with id {} not found", id));
-    }
-    return it->second->empty();
-}
+    std::copy_if(
+      previous.begin(),
+      previous.end(),
+      std::back_inserter(to_remove),
+      [&current](const model::broker_shard& prev_bs) {
+          auto it = std::find(current.begin(), current.end(), prev_bs);
+          return it == current.end();
+      });
 
-void partition_allocator::test_only_saturate_all_machines() {
-    for (auto& [id, m] : _machines) {
-        m->_partition_capacity = 0;
-        for (auto& w : m->_weights) {
-            w = allocation_node::max_allocations_per_core;
-        }
+    _state->apply_update(to_add, raft::group_id{});
```

Comment:
empty group id?

Reply:
the whole idea behind empty id is not to update highest group id stored in allocator.